File: Day #537/solution.js
// Collatz: iterate n->n/2 (even) or 3n+1 (odd) counting steps to 1; verify a range.
// Bonus longest under 1e6 via memoized step counts. Time ~O(LIMIT*avg-chain), space O(LIMIT).
'use strict';

function steps(n) {
  let c = 0;
  while (n !== 1) { n = (n % 2 === 0) ? n / 2 : 3 * n + 1; c++; }
  return c;
}

function main() {
  let allReach = true;
  for (let n = 1; n <= 20; n++) if (steps(n) < 0) allReach = false;
  console.log("Collatz conjecture holds for 1..20: " + allReach);

  const LIMIT = 1000000;
  const dp = new Int32Array(LIMIT + 1);
  let bestN = 1, bestLen = 0;
  for (let i = 2; i <= LIMIT; i++) {
    let n = i, c = 0;
    while (n !== 1 && (n > LIMIT || dp[n] === 0)) {
      n = (n % 2 === 0) ? n / 2 : 3 * n + 1; c++;
    }
    c += (n === 1) ? 0 : dp[n];
    dp[i] = c;
    if (c > bestLen) { bestLen = c; bestN = i; }
  }
  console.log(`Longest under 1000000: n=${bestN} with ${bestLen} steps`);
}

main();

File: Day #537/solution.py
# Day 537: Collatz: iterate n->n/2 (even) or 3n+1 (odd) counting steps to 1; verify a range.
# Bonus longest under 1e6 via memoized step counts. Time ~O(LIMIT*avg-chain), space O(LIMIT).

def steps(n):
    c = 0
    while n != 1:
        n = n // 2 if n % 2 == 0 else 3 * n + 1
        c += 1
    return c

def main():
    all_reach = all(steps(n) >= 0 for n in range(1, 21))
    print("Collatz conjecture holds for 1..20: " + str(all_reach).lower())

    LIMIT = 1000000
    dp = [0] * (LIMIT + 1)
    best_n, best_len = 1, 0
    for i in range(2, LIMIT + 1):
        n, c = i, 0
        while n != 1 and (n > LIMIT or dp[n] == 0):
            n = n // 2 if n % 2 == 0 else 3 * n + 1
            c += 1
        c += 0 if n == 1 else dp[n]
        dp[i] = c
        if c > best_len:
            best_len, best_n = c, i
    print(f"Longest under 1000000: n={best_n} with {best_len} steps")

if __name__ == "__main__":
    main()
